File: lib/src/main/java/com/unity3d/ads/webview/WebView.java
package com.unity3d.ads.webview;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.webkit.ValueCallback;
import android.webkit.WebSettings;

import com.unity3d.ads.log.DeviceLog;
import com.unity3d.ads.misc.Utilities;
import com.unity3d.ads.misc.ViewUtilities;
import com.unity3d.ads.webview.bridge.WebViewBridgeInterface;

import java.lang.reflect.Method;

public class WebView extends android.webkit.WebView {

	private static Method _evaluateJavascript = null;

	public WebView(Context context) {
		super(context);
		WebSettings settings = getSettings();

		if(Build.VERSION.SDK_INT >= 16) {
			settings.setAllowFileAccessFromFileURLs(true);
			settings.setAllowUniversalAccessFromFileURLs(true);
		}

		if (Build.VERSION.SDK_INT >= 19) {
			try {
				_evaluateJavascript = android.webkit.WebView.class.getMethod("evaluateJavascript", String.class, ValueCallback.class);
			} catch(NoSuchMethodException e) {
				DeviceLog.exception("Method evaluateJavascript not found", e);
				_evaluateJavascript = null;
			}
		}

		settings.setAppCacheEnabled(false);
		settings.setBlockNetworkImage(false);
		settings.setBlockNetworkLoads(false);
		settings.setBuiltInZoomControls(false);
		settings.setCacheMode(WebSettings.LOAD_NO_CACHE);
		settings.setDatabaseEnabled(false);

		if(Build.VERSION.SDK_INT >= 11) {
			settings.setDisplayZoomControls(false);
		}

		settings.setDomStorageEnabled(false);

		if(Build.VERSION.SDK_INT >= 11) {
			settings.setEnableSmoothTransition(false);
		}

		settings.setGeolocationEnabled(false);
		settings.setJavaScriptCanOpenWindowsAutomatically(false);
		settings.setJavaScriptEnabled(true);
		settings.setLightTouchEnabled(false);
		settings.setLoadWithOverviewMode(false);
		settings.setLoadsImagesAutomatically(true);

		if(Build.VERSION.SDK_INT >= 17) {
			settings.setMediaPlaybackRequiresUserGesture(false);
		}

		if(Build.VERSION.SDK_INT >= 21) {
			settings.setMixedContentMode(WebSettings.MIXED_CONTENT_NEVER_ALLOW);
		}

		settings.setNeedInitialFocus(true);
		settings.setPluginState(WebSettings.PluginState.OFF);
		settings.setRenderPriority(WebSettings.RenderPriority.NORMAL);
		settings.setSaveFormData(false);
		settings.setSavePassword(false);
		settings.setSupportMultipleWindows(false);
		settings.setSupportZoom(false);
		settings.setUseWideViewPort(true);

		setHorizontalScrollBarEnabled(false);
		setVerticalScrollBarEnabled(false);
		setInitialScale(0);
		setBackgroundColor(Color.TRANSPARENT);
		ViewUtilities.setBackground(this, new ColorDrawable(Color.TRANSPARENT));
		setBackgroundResource(0);

		addJavascriptInterface(new WebViewBridgeInterface(), "webviewbridge");
	}

	public void invokeJavascript(String data) {
		Utilities.runOnUiThread(new JavaScriptInvocation(data, this));
	}

	@Override
	public void loadUrl (String url) {
		DeviceLog.debug("Loading url: " + url);
		super.loadUrl(url);
	}

	private class JavaScriptInvocation implements Runnable {
		private String _jsString = null;
		private android.webkit.WebView _webView = null;

		public JavaScriptInvocation(String jsString, android.webkit.WebView webView) {
			_jsString = jsString;
			_webView = webView;
		}

		@Override
		public void run() {
			if (_jsString != null) {
				try {
					if (Build.VERSION.SDK_INT >= 19) {
						_evaluateJavascript.invoke(_webView, _jsString, null);
					} else {
						loadUrl(_jsString);
					}
				} catch (Exception e) {
					DeviceLog.exception("Error while processing JavaScriptString", e);
				}
			} else {
				DeviceLog.error("Could not process JavaScript, the string is NULL");
			}
		}
	}
}

File: unity-ads/src/main/java/com/unity3d/services/core/webview/WebView.kt
package com.unity3d.services.core.webview

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.webkit.ValueCallback
import android.webkit.WebSettings
import android.webkit.WebView
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import com.unity3d.services.core.configuration.Experiments
import com.unity3d.services.core.configuration.IExperiments
import com.unity3d.services.core.di.IServiceComponent
import com.unity3d.services.core.di.inject
import com.unity3d.services.core.log.DeviceLog
import com.unity3d.services.core.misc.Utilities
import com.unity3d.services.core.misc.ViewUtilities
import com.unity3d.services.core.request.metrics.SDKMetricsSender
import com.unity3d.services.core.webview.bridge.IInvocationCallbackInvoker
import com.unity3d.services.core.webview.bridge.IWebViewBridge
import com.unity3d.services.core.webview.bridge.SharedInstances
import com.unity3d.services.core.webview.bridge.WebViewBridgeInterface

open class WebView @JvmOverloads constructor(
    context: Context,
    shouldNotRequireGesturePlayback: Boolean = false,
    webViewBridge: IWebViewBridge = SharedInstances.webViewBridge,
    callbackInvoker: IInvocationCallbackInvoker = SharedInstances.webViewAppInvocationCallbackInvoker,
    experiments: IExperiments = Experiments(),
) : WebView(context), IServiceComponent {
    private val sdkMetricsSender by inject<SDKMetricsSender>()
    private val webViewBridgeInterface = WebViewBridgeInterface(webViewBridge, callbackInvoker)

    init {
        with(settings) {
            allowFileAccessFromFileURLs = true
            allowUniversalAccessFromFileURLs = true
            allowFileAccess = true
            blockNetworkImage = false
            blockNetworkLoads = false
            builtInZoomControls = false
            cacheMode = WebSettings.LOAD_NO_CACHE
            databaseEnabled = false
            displayZoomControls = false
            domStorageEnabled = false
            setEnableSmoothTransition(false)
            setGeolocationEnabled(false)
            javaScriptCanOpenWindowsAutomatically = false
            javaScriptEnabled = true
            lightTouchEnabled = false
            loadWithOverviewMode = false
            loadsImagesAutomatically = true
            mediaPlaybackRequiresUserGesture = false
            if (Build.VERSION.SDK_INT >= 21) {
                mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW
            }
            setNeedInitialFocus(true)
            pluginState = WebSettings.PluginState.OFF
            setRenderPriority(WebSettings.RenderPriority.NORMAL)
            saveFormData = false
            savePassword = false
            setSupportMultipleWindows(false)
            setSupportZoom(false)
            useWideViewPort = true
            mediaPlaybackRequiresUserGesture = !shouldNotRequireGesturePlayback
        }

        isHorizontalScrollBarEnabled = false
        isVerticalScrollBarEnabled = false
        setInitialScale(0)
        setBackgroundColor(Color.TRANSPARENT)
        ViewUtilities.setBackground(this, ColorDrawable(Color.TRANSPARENT))
        setBackgroundResource(0)

        val isWebMessageEnabled = experiments.isWebMessageEnabled
        if (isWebMessageEnabled) {
            sdkMetricsSender.sendMetric(webMessageListenerEnabledMetric())
        } else {
            sdkMetricsSender.sendMetric(webMessageListenerDisabledMetric())
        }

        val isWebMessageListenerSupported = WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)
        if (isWebMessageListenerSupported) {
            sdkMetricsSender.sendMetric(webMessageListenerSupportedMetric())
        } else {
            sdkMetricsSender.sendMetric(webMessageListenerUnsupportedMetric())
        }

        if (isWebMessageEnabled && isWebMessageListenerSupported) {
            // todo: improve implementation by using set of origins
            WebViewCompat.addWebMessageListener(this, "handleInvocation", setOf("*"), webViewBridgeInterface::onHandleInvocation)
            WebViewCompat.addWebMessageListener(this, "handleCallback", setOf("*"), webViewBridgeInterface::onHandleCallback)
        } else {
            addJavascriptInterface(webViewBridgeInterface, "webviewbridge")
        }
    }

    override fun evaluateJavascript(script: String, resultCallback: ValueCallback<String>?) {
        Utilities.runOnUiThread {
            super.evaluateJavascript(script, resultCallback)
        }
    }

    override fun loadUrl(url: String) {
        DeviceLog.debug("Loading url: $url")
        super.loadUrl(url)
    }
}
